aiida/tools/graph/age_entities.py

```diff
@@ -0,0 +1,508 @@
+# -*- coding: utf-8 -*-
+###########################################################################
+# Copyright (c), The AiiDA team. All rights reserved.                     #
+# This file is part of the AiiDA code.                                    #
+#                                                                         #
+# The code is hosted on GitHub at https://github.com/aiidateam/aiida-core #
+# For further information on the license, see the LICENSE.txt file        #
+# For further information please visit http://www.aiida.net               #
+###########################################################################
+# pylint: disable=unnecessary-pass
+"""Entities for the AiiDA Graph Explorer utility"""
+
+from abc import ABCMeta, abstractmethod
+from collections import namedtuple
+
+from aiida.orm import Node, Group
+from aiida.orm.querybuilder import QueryBuilder
+from aiida.orm.utils.links import LinkQuadruple
+
+# In order to check valid classes that are implemented:
+VALID_ENTITY_CLASSES = (Node, Group)
+
+# Edges are defined and returned as namedtuples
+GroupNodeEdge = namedtuple('GroupNodeEdge', ['node_id', 'group_id'])
+
+
+class AbstractSetContainer(metaclass=ABCMeta):
+    """This class defines a set that can be speciaized to contain either nodes (aiida nodes )"""
+
+    @abstractmethod
+    def _check_self_and_other(self, other):
+        """
+        Utility function. When called, will check whether self and other instance
+        are compatible. Same aiida classes, same identifiers, length of tuples...
+        """
+        pass
+
+    @abstractmethod
+    def _check_input_for_set(self, input_for_set):
+        """
+        When giving me something to the set, this utility function can be used
+        to do the right thing.
+        """
+        pass
+
+    @abstractmethod
+    def copy(self):
+        """
+        Copy with all data
+        """
+        pass
+
+    @abstractmethod
+    def get_template(self):
+        """
+        Copy the template
+        """
+        pass
+
+    @property
+    def keyset(self):
+        """Set containing the keys of the entities"""
+        return self._keyset
+
+    @keyset.setter
+    def keyset(self, inpset):
+        """Use with care: will not check if the keys are consistent ids"""
```

Comment:
Ok, so basically this means that if you give it id numbers it has no way of knowing if these are all nodes or all groups or even if they exist and will just trust the input, so you could potentially end up mixing ids of nodes with ids of groups. Afterwards when you work with the class it will make checks (if you try to add two `SetContainers` it checks if they are the same kind or if you are trying to add a container for nodes with a container for groups for example). The reason it doesn't check is that it would then require something like doing a query for every single id you add every time you create or copy one of these containers, which would make the whole thing rather unusable.

Reply:
Maybe add some comment mentioning the checks are performed upstream in code, and that there is no checks because there is no way of matching ID with type.